### PR TITLE
xxd: allow printing bits low order first with -L

### PR DESCRIPTION
**Pull request by afontenot**
Hi, 

Conventionally when printing a hex dump each byte is printed as an ordinary pair of hexadecimal numbers for easy reading. This makes sense, since for example 0x61 is the ASCII 'a'. Sometimes however one needs a binary representation, for instance when debugging bitfields. Currently in this case xxd will print 01100001, the binary representation of 0x61. This also makes some sense. Most of us read from left to right and since bits aren't usually individually addressable, simply showing the binary representation of the whole byte is a reasonable default.

However, there is some reason to want to print the bits in the opposite order. On modern little endian systems, [the bits are stored](https://www.linuxjournal.com/article/6788) in the CPU in "little endian" bit order, or low order first. (That is, the least significant bit in the least significant byte is the one that comes first.) Likewise, even though the C standard says that the layout of structs is an implementation detail, the well-known compilers also appear to store data low order first on little endian systems. [GCC for example.](https://gcc.gnu.org/ml/gcc/2004-09/msg00527.html) Lastly, some networking hardware also sends data over the wire with the low order bit coming first.

I've occasionally encountered situations where it would be nice to see the way some bit of data *really* looks on the disk or in memory - e.g. when trying to parse an obscure format with bitfields that stretch across multiple bytes. Plus I admit reading little endian hex dumps with the bits printed in big endian order throws me off at times. I couldn't find any open source tools that would do this.

So in short, I added support to xxd. Thanks for reading - if you're interested in pulling this I'm happy to make changes as needed.

Sample output:

    % ./xxd -b endian.txt
    00000000: 00100010 00111000 01001110 00001010                    "8N.
    % ./xxd -L endian.txt
    00000000: 01000100 00011100 01110010 01010000                    "8N.

### CODE REVIEW

**Comment by brammool** [comment]
> Conventionally when printing a hex dump each byte is printed as an
> ordinary pair of hexadecimal numbers for easy reading. This makes
> sense, since for example 0x61 is the ASCII 'a'. Sometimes however one
> needs a binary representation, for instance when debugging bitfields.
> Currently in this case xxd will print 01100001, the binary
> representation of 0x61. This also makes some sense. Most of us read
> from left to right and since bits aren't usually individually
> addressable, simply showing the binary representation of the whole
> byte is a reasonable default.
> 
> However, there is some reason to want to print the bits in the
> opposite order. On modern little endian systems, [the bits are
> stored](https://www.linuxjournal.com/article/6788) in the CPU in
> "little endian" bit order, or low order first. (That is, the least
> significant bit in the least significant byte is the one that comes
> first.) Likewise, even though the C standard says that the layout of
> structs is an implementation detail, the well-known compilers also
> appear to store data low order first on little endian systems. [GCC
> for example.](https://gcc.gnu.org/ml/gcc/2004-09/msg00527.html)
> Lastly, some networking hardware also sends data over the wire with
> the low order bit coming first.
> 
> I've occasionally encountered situations where it would be nice to see
> the way some bit of data *really* looks on the disk or in memory -
> e.g. when trying to parse an obscure format with bitfields that
> stretch across multiple bytes. Plus I admit reading little endian hex
> dumps with the bits printed in big endian order throws me off at
> times. I couldn't find any open source tools that would do this.
> 
> So in short, I added support to xxd. Thanks for reading - if you're
> interested in pulling this I'm happy to make changes as needed.
> 
> Sample output:
> 
>     % ./xxd -b endian.txt
>     00000000: 00100010 00111000 01001110 00001010                    "8N.
>     % ./xxd -L endian.txt
>     00000000: 01000100 00011100 01110010 01010000                    "8N.

I would argue this could also be done with a bit of Vim scripting, but
since you wrote the patch anyway, might as well include it.

Please add a test case in src/testdir/test_xxd.vim

-- 
Article in the first Free Software Magazine: "Bram Moolenaar studied
electrical engineering at the Technical University of Delft and
graduated in 1985 on a multi-processor Unix architecture."
Response by "dimator": Could the school not afford a proper
stage for the ceremony?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
No reply for months, therefore closing.  Please reopen when a test was added.
